core/src/main/scalajvm/scalapb/zio_grpc/server/ZServerCallHandler.scala

```diff
@@ -108,30 +111,62 @@ object ZServerCallHandler {
-          case Exit.Success(res)   => call.sendMessage(res).as(true)
-          case Exit.Failure(cause) =>
-            cause.failureOrCause match {
-              case Left(Some(status)) => ZIO.fail(status)
-              case Left(None)         => ZIO.succeed(false)
-              case Right(cause)       => ZIO.failCause(cause)
+    def takeFromQueue(queue: Dequeue[Exit[Option[Status], Res]]): ZIO[Any, Status, Unit] =
+      queue.takeAll.flatMap(takeFromCache(_, queue))
+
+    def takeFromCache(
+        xs: Chunk[Exit[Option[Status], Res]],
+        queue: Dequeue[Exit[Option[Status], Res]]
+    ): ZIO[Any, Status, Unit] =
+      ZIO.suspendSucceed {
+        @tailrec def innerLoop(loop: Boolean, i: Int): IO[Status, Unit] =
+          if (loop) {
+            xs(i) match {
+              case Failure(cause) =>
+                cause.failureOrCause match {
+                  case Left(Some(status)) =>
+                    ZIO.fail(status)
+                  case Left(None)         =>
+                    ZIO.unit
+                  case Right(cause)       =>
+                    ZIO.failCause(cause)
+                }
+              case Success(value) =>
+                call.call.sendMessage(value)
+                // at this point we have consumed i + 1 elements from the chunk
+                val newI = i + 1
+                // the loop iteration may only continue if the call can
+                // still accept elements and we have more elements to send
+                innerLoop(call.call.isReady && newI < xs.length, newI)
             }
-        }
-        .repeatWhileZIO(res => call.isReady.map(_ && res))
+          } else if (call.call.isReady)
+            // ^ if we reached the end of the chunk but the call can still
+            // proceed, we pull from the queue and continue
+            takeFromQueue(queue)
+          else
+            // ^ otherwise, we wait for the call to be ready and then start again
+            outerLoop(Some(xs.drop(i)), queue)
+
+        if (xs.isEmpty)
+          takeFromQueue(queue)
+        else
+          innerLoop(true, 0)
+      }
 
-    def outerLoop(queue: Dequeue[Exit[Option[Status], Res]]): ZIO[Any, Status, Boolean] =
-      (call.awaitReady *> innerLoop(queue))
-        .repeatWhile(identity)
+    def outerLoop(
+        cache: Option[Chunk[Exit[Option[Status], Res]]],
+        queue: Dequeue[Exit[Option[Status], Res]]
+    ): ZIO[Any, Status, Unit] =
+      (call.awaitReady *> cache.fold(takeFromQueue(queue))(takeFromCache(_, queue)))
 
     for {
       queueSize <- backpressureQueueSize
-      _         <- ZIO.scoped[Any](
-                     stream
-                       .toQueueOfElements(queueSize)
-                       .flatMap(outerLoop)
-                   )
+      _         <- ZIO
+                     .scoped[Any](
+                       stream
+                         .toQueueOfElements(queueSize)
+                         .flatMap(outerLoop(None, _).fork.flatMap(_.join))
```

Comment:
can you add comment on why `fork.flatMap(_.join)` ?

Reply:
After re-reading the implementation of `toQueueOfElements`, the producer is forked so this is not necessary :facepalm:

Reply:
Undone